File: go/broho-push.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"

	"broho-push.go/url_broker"

	"github.com/kirsle/configdir"
	"google.golang.org/grpc"
)

type Config struct {
	BindAddress string `json:"bind-address"`
	BindPort    string `json:"bind-port"`
}

func load_config() Config {
	configPath := configdir.LocalConfig("broho")
	err := configdir.MakePath(configPath) // Ensure it exists.
	if err != nil {
		panic(err)
	}

	// Deal with a JSON configuration file in that folder.
	configFile := filepath.Join(configPath, "config.json")
	var config Config

	// Does the file not exist?
	if _, err = os.Stat(configFile); os.IsNotExist(err) {
		// Create the new config file.
		config = Config{"127.0.0.1", "50051"}
		fh, err := os.Create(configFile)
		if err != nil {
			panic(err)
		}
		defer fh.Close()

		encoder := json.NewEncoder(fh)
		encoder.Encode(&config)
	} else {
		// Load the existing file.
		fh, err := os.Open(configFile)
		if err != nil {
			panic(err)
		}
		defer fh.Close()

		decoder := json.NewDecoder(fh)
		decoder.Decode(&config)
	}
	return config
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println(filepath.Base(os.Args[0]), "requires at least one URL as argument")
		os.Exit(1)
	} else {
		var config Config = load_config()
		var endpoint string = config.BindAddress + ":" + config.BindPort

		var channel *grpc.ClientConn
		channel, err := grpc.Dial(endpoint, grpc.WithInsecure())
		if err != nil {
			fmt.Printf("Could not connect: %s\n", err)
		}
		defer channel.Close()

		client := url_broker.NewURLBrokerServiceClient(channel)
		re := regexp.MustCompile(`https?://.*`)

		for _, arg := range os.Args[1:] {
			if re.MatchString(arg) {
				fmt.Printf("Pushing URL %s\n", arg)
				response, err := client.PushURL(context.Background(), &url_broker.URL{Url: arg})
				if err != nil {
					fmt.Printf("Error occurred when pushing URL %s: %s\n", arg, err)
				} else if response.ResponseCode != 0 {
					fmt.Printf("Error occurred when pushing URL %s (response code: %d)\n", arg, response.ResponseCode)
				}
			} else {
				fmt.Printf("Invalid URL: %s\n", arg)
			}
		}
	}
}

func URLBrokerServiceClient(channel *grpc.ClientConn) {
	panic("unimplemented")
}

File: python/broho-push.py
#!/usr/bin/env python3

import logging
import argparse
import re
import configparser
import os
import sys
import grpc
from protos.url_broker_pb2 import URL
from protos.url_broker_pb2_grpc import URLBrokerServiceStub

logger = logging.getLogger()
logger.addHandler(logging.StreamHandler(sys.stdout))
logger.setLevel(logging.INFO)

def load_config():
  config = configparser.ConfigParser()
  standard_config_file_locations = [ "/etc/broho/broho.conf",
                                     "~/.config/broho/broho.conf" ]
  for config_file in standard_config_file_locations:
    if os.path.isfile(config_file):
      logger.debug(f"Loaded config file at {config_file}")
      try:
        config.read_file(open(config_file))
        if 'Global' in config.sections():
          return config
      except PermissionError:
        logger.error(f"Couldn't open config file '{config_file}' (permission denied)")
  
  logger.info("No config file found (using defaults)")
  config['Global'] = { 'BindAddress': '127.0.0.1',
                       'BindPort': '50051'
  }
  return config

def main():
  parser = argparse.ArgumentParser()
  parser.add_argument('urls', nargs='+', help='URL(s) to open on host')
  args = parser.parse_args()
  
  config = load_config()
  bind_address = config['Global'].get('BindAddress', '127.0.0.1')
  bind_port = config['Global'].get('BindPort', '50051')
  endpoint = f"{bind_address}:{bind_port}"

  try:
    channel = grpc.insecure_channel(endpoint)
    client = URLBrokerServiceStub(channel)
    for url_string in args.urls:
      if re.match(r'https?://.*', url_string):
        url = URL(url=url_string)
        logger.info(f"Pushing URL {url_string}")
        response = client.PushURL(url)
        if response.response_code != 0:
          logger.error(f"Error occurred when pushing URL {url.url} (response code: {response.response_code}")
      else:
        logger.error(f"Invalid URL: {url_string}")
  except grpc.RpcError as err:
    logger.error(err.details())

# Handle interruption
if __name__ == "__main__":
  try:
    main()
  except KeyboardInterrupt:
    pass
